### PR TITLE
Get Global Tail From StreamFiles

### PR DESCRIPTION
**Pull request by Maithem**

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **716f2bb**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #560 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/Q3yOYe0OUXaLpNVt74igXRJhYNS6sQ7W)

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java [hunk 1]
```diff
@@ -93,4 +95,16 @@ public synchronized void setServerEpoch(long serverEpoch) {
         //TODO need to figure out if we can remove this redundancy
         serverRouter.setServerEpoch(serverEpoch);
     }
+
+    public synchronized long getTailSegment() {
```
Comment (medhavidhawan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java]: why do you need this synchronized. I remember the datastore gave you that. I might be wrong.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java [hunk 2]
```diff
@@ -93,4 +95,16 @@ public synchronized void setServerEpoch(long serverEpoch) {
...
+        Long tailSegment = dataStore.get(Long.class, PREFIX_TAIL_SEGMENT, KEY_TAIL_SEGMENT);
+        return tailSegment == null ? 0 : tailSegment;
+    }
+
+    public synchronized void setTailSegment(long tailSegment) {
```
Comment (medhavidhawan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java]: not sure if you need to synchronize again. Please also add comments.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java [hunk 3]
```diff
@@ -93,4 +95,16 @@ public synchronized void setServerEpoch(long serverEpoch) {
...
+        dataStore.put(Long.class, PREFIX_TAIL_SEGMENT, KEY_TAIL_SEGMENT, tailSegment);
+        // Set the epoch in the router as well.
+        //TODO need to figure out if we can remove this redundancy
+        serverRouter.setServerEpoch(tailSegment);
```
Comment (medhavidhawan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java]: what is this ?

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 4]
```diff
@@ -84,14 +86,46 @@
     private Map<String, SegmentHandle> writeChannels;
     private Set<FileChannel> channelsToSync;
     private MultiReadWriteLock segmentLocks = new MultiReadWriteLock();
+    final private ServerContext serverContext;
+    final private AtomicLong globalTail = new AtomicLong(0L);
+    final private AtomicLong tempGlobalTail = new AtomicLong(0L);
```
Comment (medhavidhawan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: this name is not very intuitive. please change it to something better. Also please add comments for these fields.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java (hunk 3 above)
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java]: Copy paste error ?

**Comment by Maithem** [comment]
@zalokhan @medhavidhawan This is just a stash, thats why i didnt add you for review. I'll ping you tomorrow when its ready

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **d7cda3f**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #560 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/mI83vtvyv5aWgc34mSQUDJwwJJLdhrOC)

**Pull request merged**